lnd.go

```diff
@@ -630,105 +633,203 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+	// Genereate a TLS pair if we don't have one yet.
+	err := t.createTLSPair()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certData, parsedCert, err := cert.LoadCert(
+		t.cfg.TLSCertPath, t.cfg.TLSKeyPath,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check to see if the certificate needs to be renewed. If it does, we
+	// return the newly generated certificate data instead.
+	certData, err = t.certMaintenance(parsedCert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tlsCfg := cert.TLSConfFromCert(certData)
+
+	cleanUp := t.setUpLetsEncrypt(&certData, tlsCfg)
+
+	return tlsCfg, cleanUp, nil
+}
+
+// If a TLS pair doesn't exist yet, create them and write them to disk.
+func (t *TLSManager) createTLSPair() error {
 	// Ensure we create TLS key and certificate if they don't exist.
-	if !fileExists(cfg.TLSCertPath) && !fileExists(cfg.TLSKeyPath) {
+	if !fileExists(t.cfg.TLSCertPath) && !fileExists(t.cfg.TLSKeyPath) {
 		rpcsLog.Infof("Generating TLS certificates...")
 		err := cert.GenCertPair(
-			"lnd autogenerated cert", cfg.TLSCertPath,
-			cfg.TLSKeyPath, cfg.TLSExtraIPs, cfg.TLSExtraDomains,
-			cfg.TLSDisableAutofill, cfg.TLSCertDuration,
+			"lnd autogenerated cert", t.cfg.TLSCertPath,
+			t.cfg.TLSKeyPath, t.cfg.TLSExtraIPs,
+			t.cfg.TLSExtraDomains, t.cfg.TLSDisableAutofill,
+			t.cfg.TLSCertDuration,
 		)
 		if err != nil {
-			return nil, nil, nil, nil, err
+			return err
 		}
 		rpcsLog.Infof("Done generating TLS certificates")
 	}
 
-	certData, parsedCert, err := cert.LoadCert(
-		cfg.TLSCertPath, cfg.TLSKeyPath,
-	)
-	if err != nil {
-		return nil, nil, nil, nil, err
-	}
+	return nil
+}
+
+// certMaintenance checks if the certificate IP and domains matches the config,
+// and renews the certificate if either this data is outdated or the
+// certificate is expired.
+func (t *TLSManager) certMaintenance(parsedCert *x509.Certificate,
+) (tls.Certificate, error) {
 
 	// We check whether the certificate we have on disk match the IPs and
 	// domains specified by the config. If the extra IPs or domains have
 	// changed from when the certificate was created, we will refresh the
 	// certificate if auto refresh is active.
 	refresh := false
-	if cfg.TLSAutoRefresh {
+	var err error
+	if t.cfg.TLSAutoRefresh {
 		refresh, err = cert.IsOutdated(
-			parsedCert, cfg.TLSExtraIPs,
-			cfg.TLSExtraDomains, cfg.TLSDisableAutofill,
+			parsedCert, t.cfg.TLSExtraIPs,
+			t.cfg.TLSExtraDomains, t.cfg.TLSDisableAutofill,
 		)
 		if err != nil {
-			return nil, nil, nil, nil, err
+			return tls.Certificate{}, err
 		}
 	}
 
+	var reloadedCertData tls.Certificate
 	// If the certificate expired or it was outdated, delete it and the TLS
 	// key and generate a new pair.
 	if time.Now().After(parsedCert.NotAfter) || refresh {
 		ltndLog.Info("TLS certificate is expired or outdated, " +
 			"generating a new one")
 
-		err := os.Remove(cfg.TLSCertPath)
+		err := os.Remove(t.cfg.TLSCertPath)
```

Comment:
Same in this `if`, now that we're in a method, we can just return early if there's nothing to do and have the refresh logic non-indented.